File: packages/services/recipe-workers/src/common/observability.ts
/**
 * Sentry for the recipe-workers Lambdas (plan U16/U22).
 *
 * ⛔ WHY IT WAS MISSING AND WHY THAT MATTERED. These ten functions wrote structured lines to stdout and
 * nothing else. Until plan U15 their log group had no drain at all, so a thrown error reached CloudWatch and
 * stopped — no issue, no grouping, no stack trace anybody would see, no alert. The drain fixes visibility of
 * LOGS; it cannot turn a failure into an issue, because a forwarded log line is a log line.
 *
 * ⛔ AND THE TWO PATHS MUST NOT DOUBLE-COUNT. The drain forwards stdout, so an error written to stdout AND
 * captured by the SDK is one failure recorded twice — once as an issue, once as an ERROR-severity log — and
 * Sentry's alerting counts them separately. So when Sentry is initialised, `logger.error` goes to Sentry's
 * logger INSTEAD of stdout. `info` and `warn` stay on stdout, because CloudWatch is the durable record and
 * neither creates an issue.
 *
 * ⚠️ EMF METRIC LINES ARE UNTOUCHED. `metrics.ts` writes them straight to stdout and the drain's filter
 * pattern excludes `_aws`, so they reach CloudWatch as metrics and never Sentry as logs. Routing them
 * through this module would turn a metric into a log line and lose the metric.
 *
 * ⚠️ WITH NO DSN THIS IS INERT. Every local run, every unit test and every stage that has not had its
 * parameter written behaves exactly as before — `initObservability` returns `false` and `withObservability`
 * hands back the handler it was given.
 */
import * as Sentry from '@sentry/aws-serverless';
import type { Context } from 'aws-lambda';
import { scrubEvent, scrubLog } from '@kitchensink/observability-scrubbers';

/** Whether `Sentry.init` ran. Read by {@link isObservabilityEnabled} and by the logger below. */
let initialised = false;

/**
 * Initialise Sentry from the environment, once.
 *
 * ⚠️ A FUNCTION, not a module side effect, for the reason `identity-webhooks` learned: a top-level
 * `Sentry.init` runs during Lambda INIT, where a construction failure produces no structured log and no
 * metric — the invocation simply never starts and the only evidence is an absent log stream.
 *
 * @param env - The environment to read; injectable so the suite can drive both branches.
 * @returns Whether Sentry was initialised.
 * @sideEffect Configures the global Sentry client.
 */
export function initObservability(env: NodeJS.ProcessEnv = process.env): boolean {
    const dsn = env['SENTRY_DSN'];

    if (!dsn || initialised) {
        return initialised;
    }

    Sentry.init({
        dsn,
        environment: env['STAGE'] ?? 'dev',
        // ⚠️ Spread, not `release: undefined`: a release named "undefined" groups unrelated deploys, which
        // is harder to notice than no release at all.
        ...(env['SENTRY_RELEASE'] === undefined ? {} : { release: env['SENTRY_RELEASE'] }),
        tracesSampleRate: Number(env['SENTRY_TRACES_SAMPLE_RATE'] ?? '0'),
        enableLogs: true,
        sendDefaultPii: false,
        beforeSend: scrubEvent,
        beforeSendLog: scrubLog,
    });

    initialised = true;

    return true;
}

/** Whether Sentry is live in this process. */
export function isObservabilityEnabled(): boolean {
    return initialised;
}

let isColdStart = true;

/**
 * Wrap a Lambda handler so a thrown error becomes a Sentry ISSUE, with per-invocation context.
 *
 * ⛔ An ISSUE, not a log line. That is the whole difference this module makes: the log drain (U15) already
 * carries the stdout line, and a forwarded log line is a log line — it does not group, it does not carry a
 * stack trace Sentry can symbolicate, and nothing alerts on it. Errors and logs are different products.
 *
 * @param handler - The Lambda handler.
 * @returns The wrapped handler, or the original one when Sentry is not initialised.
 * @sideEffect Installs Sentry instrumentation around the handler.
 */
export function withObservability<F extends (...args: never[]) => unknown>(handler: F): F {
    if (!initialised) {
        return handler;
    }

    const instrumented = ((...args: unknown[]): unknown => {
        const context = args[1] as Context | undefined;
        const coldStart = isColdStart;
        isColdStart = false;

        // ⚠️ Every attribute is OPTIONAL-safe. Half these functions are SCHEDULED — invoked by EventBridge
        // with no SQS event and, in the unit suite, called with no arguments at all — so reading
        // `context.awsRequestId` unguarded would turn "Sentry is enabled" into "every scheduled handler
        // throws on entry", which is the worst possible way for an observability change to fail.
        Sentry.getIsolationScope().setAttributes({
            aws_request_id: context?.awsRequestId ?? 'none',
            cold_start: coldStart,
            function_name: context?.functionName ?? 'unknown',
            function_version: context?.functionVersion ?? 'unknown',
            serviceName: 'recipe-workers',
        });

        return (handler as unknown as (...inner: unknown[]) => unknown)(...args);
    }) as unknown as F;

    // ⛔ The cast is to `F`, not to `Handler`, and the reason is that these ten functions do NOT share one
    // signature: the SQS consumers take `(event, context)`, the scheduled sweepers take nothing, and the
    // unit suites call the scheduled ones with no arguments. Typing this as `Handler<TEvent, TResult>` would
    // have forced every one of those call sites to invent an event and a context — editing tests to satisfy
    // a wrapper rather than because behaviour changed, which is exactly the edit `CLAUDE.md` calls the worst
    // outcome available. Preserving `F` keeps each handler's own contract intact.
    return Sentry.wrapHandler(instrumented as never) as unknown as F;
}

/**
 * Report an error line to Sentry rather than stdout.
 *
 * ⛔ INSTEAD OF, not as well as. The drain forwards stdout, so a line written in both places is one failure
 * recorded twice — once as an issue if it also throws, and once as an ERROR-severity log — and Sentry's
 * alerting counts them separately.
 *
 * @param message - The log message.
 * @param attributes - Structured attributes, scrubbed by `beforeSendLog`.
 * @returns Whether the line was taken by Sentry (so the caller knows not to write it).
 * @sideEffect Emits a Sentry log entry.
 */
export function reportErrorLine(message: string, attributes?: Record<string, unknown>): boolean {
    if (!initialised) {
        return false;
    }

    Sentry.logger.error(message, attributes);

    return true;
}

/**
 * Capture an error that is handled rather than thrown.
 *
 * ⚠️ For the paths that swallow deliberately — a metered-and-swallowed verdict write, a best-effort settle —
 * where the whole point is that the handler continues. Those are exactly the failures nothing else records.
 *
 * @param error - The caught error.
 * @param context - Identifiers describing where it happened. Never payload text.
 * @sideEffect Emits a Sentry issue.
 */
export function captureHandled(error: unknown, context: Record<string, string | number | boolean>): void {
    if (!initialised) {
        return;
    }

    Sentry.withScope((scope) => {
        scope.setContext('work_unit', context);
        Sentry.captureException(error);
    });
}

File: packages/services/ingredient-parser/src/observability.py
"""Sentry for the CRF ingredient-parse engine (plan U21).

⛔ WHY THIS EXISTS. An exception inside `parse_ingredient` is per-LINE by design: `handler.py` catches it and
reports a `status` so one bad sentence does not lose the other 199 parses in the batch. That is right for the
caller and useless for diagnosis — the caller learns "that line failed" and never learns WHY, because the
Python traceback that names the cause went to a log group nobody read. Reporting it makes the engine's own
failures visible as issues, with a stack trace, grouped by where they happen rather than by which recipe hit
them.

⚠️ INERT WITHOUT A DSN, and that is what keeps every local run and every test unchanged. `init()` returns
`False` and `report()` does nothing, so the handler behaves exactly as it did before this module existed.

⚠️ NO PII. The engine is handed ingredient phrases — a cook's own words — and `send_default_pii` is off, but
that setting governs request/user data rather than exception values. What protects the phrase here is that
this module reports EXCEPTIONS and never the input: `report()` takes a message and identifiers, and the
handler passes neither the sentence nor the batch.
"""

from __future__ import annotations

import logging
import os

_logger = logging.getLogger(__name__)
_initialised = False


def init(engine_version: str) -> bool:
    """Initialise Sentry from the environment.

    :param engine_version: The pinned engine version, reported as the release so a traceback is attributable
        to the model that produced it — which for this service is the thing that actually changes behaviour.
    :returns: Whether Sentry was initialised.
    """
    global _initialised

    if _initialised:
        return True

    dsn = os.environ.get("SENTRY_DSN")

    if not dsn:
        return False

    try:
        import sentry_sdk
    except ImportError:  # pragma: no cover - the asset guard makes this unreachable in a deploy
        # ⚠️ Swallowed DELIBERATELY, and only here. A missing SDK must never stop the parser parsing: this
        # module is observability, and an import error in observability that takes the service down is
        # strictly worse than the blindness it was added to fix.
        _logger.warning("sentry-sdk is not installed; the parser runs unreported")
        return False

    sentry_sdk.init(
        dsn=dsn,
        environment=os.environ.get("STAGE", "dev"),
        release=engine_version,
        send_default_pii=False,
        # ⚠️ Tracing OFF by default. This function is invoked in a batch by one caller that already traces;
        # a second trace of the same work would double the spans and tell nobody anything new.
        traces_sample_rate=float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0")),
    )
    _initialised = True

    return True


def report(error: BaseException, **context: object) -> None:
    """Report a handled exception, with identifiers only.

    ⚠️ ``object``, not ``Any``. ``Any`` silences the type checker for every caller — ruff's ANN401 refuses
    it for exactly that reason — while ``object`` still accepts any value and keeps the checker honest about
    what may be done with it. Nothing here does anything with a context value except hand it to Sentry.

    :param error: The caught exception.
    :param context: Identifiers describing where it happened. ⛔ NEVER the ingredient phrase — the whole
        point of reporting here is to name the failure without copying the cook's words into a place the
        erasure path does not reach.
    """
    if not _initialised:
        return

    try:
        import sentry_sdk

        with sentry_sdk.new_scope() as scope:
            scope.set_context("work_unit", context)
            sentry_sdk.capture_exception(error)
    except Exception:  # pragma: no cover - reporting must never fail the parse
        _logger.warning("could not report an engine failure to Sentry", exc_info=True)
